### PR TITLE
[MID-14] fix env credentials authenticaction

### PR DESCRIPTION
**Pull request by IvanKoshan**
https://trello.com/c/pak9hVVZ/14-add-auth-for-admin-panel

### CODE REVIEW

**Review by Yivgha — APPROVED** [review]

**Comment by IvanKoshan** [comment]
Посилання на деплой https://ivan-koshans-mayak.vercel.app/

**Review by Yivgha — APPROVED** [review]

**Review by alex-popov-tech — APPROVED** [review]

File: src/lib/auth.js
Comment (yataras) [inline src/lib/auth.js]: @IvanKoshan Please, check if 1 hour maxAge matches with business requirements.

File: src/lib/auth.js
Comment (yataras) [inline src/lib/auth.js]: Consider error handling or at least logging an error to make debugging easier

File: src/lib/auth.js
Comment (BohdanLeshchyshyn) [reply src/lib/auth.js]: @yataras I believe it does not matter for now
because auth will be used only for the admin panel.
and it is just the first version of admin panel.

Comment (yataras) [reply src/lib/auth.js]: @BohdanLeshchyshyn OK

**Review by yataras — APPROVED** [review]

File: src/app/admin/adminApp.js
Comment (BohdanLeshchyshyn) [inline src/app/admin/adminApp.js]: Why do you put await before call signIn function, if the callback function that you pass here is not async?

File: src/lib/auth.js
Comment (IvanKoshan) [reply src/lib/auth.js]: Added additional logging for authentication events, sign ins, sign outs and errors

File: src/app/admin/adminApp.js
Comment (BohdanLeshchyshyn) [reply src/app/admin/adminApp.js]: Don't you think that you can simplify this part of code just by writing - 
`login: signIn,`

File: src/app/admin/adminApp.js
Comment (BohdanLeshchyshyn) [inline src/app/admin/adminApp.js]: is it required to define checkError function?
if not,  i do not see any reason to define it and return Promise.resolve() especially if you do not put async word before function definition

File: src/app/admin/adminApp.js
Comment (BohdanLeshchyshyn) [inline src/app/admin/adminApp.js]: I believe I already saw the same string `/admin#/login`
in file 
src/lib/auth.js

It is not good to repeat the same variable several times because if you change the route to the admin panel. You will have to go through all the places and check them. And it will be easy to miss one place - better to have one variable for such repeating things.

please, create a variable for it and use it in both places.

File: src/app/admin/adminApp.js
Comment (IvanKoshan) [reply src/app/admin/adminApp.js]: The callback function is not async, however it returns a Promise (the function inside of it returns a promise). We're awaiting the singin and signout processes on the backend to finish before continuing on the frontend

Comment (BohdanLeshchyshyn) [reply src/app/admin/adminApp.js]: Ok. Leave it for now - we will simplify it later.
Please follow the second comment in this thread

File: src/app/admin/adminApp.js
Comment (IvanKoshan) [reply src/app/admin/adminApp.js]: checkError and getPermissions functions are used by react admin, and must have a definition

File: src/app/admin/adminApp.js
Comment (IvanKoshan) [reply src/app/admin/adminApp.js]: done

File: src/app/admin/adminApp.js
Comment (BohdanLeshchyshyn) [inline src/app/admin/adminApp.js:1]: Okay, Just to do it faster,
what do you think if to do a bit of refactoring?
```js
'use client';

import { getSession, signIn, signOut } from 'next-auth/react';
import {
  Admin,
  ListGuesser,
  Resource,
  ShowGuesser,
  EditGuesser,
} from 'react-admin';
import { dataProvider } from 'ra-data-simple-prisma';
import { LOGIN_URL } from '@/lib/consts';

const adminDataProvider = dataProvider('/api/admin');
const authProvider = {
  login: async (credentials) => signIn('credentials', credentials, { redirect: false }),
  logout: async () => signOut({ callbackUrl: LOGIN_URL }),
  checkAuth: async () => {
    const session = await getSession();

    return session ? Promise.resolve() : Promise.reject();
  },
  checkError: async (error) => {
    // till we have a better way to handle errors in this place
    throw error;
  },
  getIdentity: async () => {
    const session = await getSession();

    return {
      id: session.user.id,
      fullName: session.user.name,
    };
  },
  getPermissions: () => Promise.resolve(),
};

export default function AdminPage() {
  return (
    <Admin dataProvider={adminDataProvider} authProvider={authProvider}>
      <Resource
        name="Therapy"
        list={ListGuesser}
        edit={EditGuesser}
        show={ShowGuesser}
      />
      <Resource
        name="Specialist"
        list={ListGuesser}
        edit={EditGuesser}
        show={ShowGuesser}
      />
    </Admin>
  );
}

```

File: src/lib/auth.js
Comment (BohdanLeshchyshyn) [inline src/lib/auth.js:1]: And here

```js
import NextAuth from 'next-auth';
import CredentialsProvider from 'next-auth/providers/credentials';
import { z } from 'zod';
import { env } from '@/lib/env';

const adminSessionLifetime = 60 * 60; // 1 hour in seconds

export const { auth, handlers } = NextAuth({
  logger: { error: console.error },
  secret: process.env.AUTH_SECRET,
  session: {
    strategy: 'jwt',
    maxAge: adminSessionLifetime,
  },
  providers: [
    CredentialsProvider({
      async authorize(credentials) {
        const adminCheckResult = z
          .object({
            username: z.string().refine((val) => val === env.ADMIN_USERNAME),
            password: z.string().refine((val) => val === env.ADMIN_PASSWORD),
          })
          .safeParse(credentials);

        if (!adminCheckResult.success) {
          throw new Error('Invalid credentials');
        }

        return { name: adminCheckResult.data.username };
      },
    }),
  ],
});

```

File: src/app/admin/adminApp.js
Comment (IvanKoshan) [reply src/app/admin/adminApp.js:1]: Did it

File: src/lib/auth.js
Comment (IvanKoshan) [reply src/lib/auth.js:1]: did it here too

File: src/app/admin/adminApp.js
Comment (BohdanLeshchyshyn) [reply src/app/admin/adminApp.js:1]: Thanks

**Comment by BohdanLeshchyshyn** [comment]
Now let's wait till linting PR will be merged and after that, you can update your branch and check on linting errors and fix them.
Ping me after that happens and i will merge PR

**Comment by IvanKoshan** [comment]
@BohdanLeshchyshyn

**Pull request merged**
